File: code/ch19/19.3.3. 72.minDistance.py
class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        m = len(word1)
        n = len(word2)
        pre = [0] * (n + 1)
        cur = [0] * (n + 1)

        for i in range(1, n + 1):
            pre[i] = i
        for i in range(1, m + 1):
            cur[0] = i
            for j in range(1, n + 1):
                if word1[i - 1] == word2[j - 1]:
                    cur[j] = pre[j - 1]
                else:
                    cur[j] = min(pre[j], pre[j - 1], cur[j - 1]) + 1
            # move on
            pre = cur.copy()
        # 最后进行了一次交换， cur变成了pre，因此我们应该取pre[n]，而不是cur[n]
        return pre[n]

File: code/ch19/19.3.3. 72.minDistance.cpp
#include <string>
#include <cstdlib>
#include <algorithm>
using namespace std;

class Solution {
public:
  int minDistance(string word1, string word2) {
    int m = word1.length(), n = word2.length();
    int *pre = (int *)malloc((n + 1) * sizeof(int));
    int *cur = (int *)malloc((n + 1) * sizeof(int));
    for (int i = 0; i <= n; i++) pre[i] = cur[i] = 0;
    for (int i = 1; i <= n; i++) pre[i] = i;
    for (int i = 1; i <= m; i++) {
      cur[0] = i;
      for (int j = 1; j <= n; j++) {
        if (word1[i - 1] == word2[j - 1])
          cur[j] = pre[j - 1];
        else
          cur[j] = min(min(pre[j], pre[j - 1]),
              cur[j - 1]) + 1;
      }
      for (int k = 0; k <= n; k++) pre[k] = cur[k];
    }
    int ans = pre[n];
    free(pre);
    free(cur);
    return ans;
  }
};
